internal/p2p/broadcast/reliable.go

```diff
@@ -0,0 +1,355 @@
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+const (
+	// TODO: enough or too much?
+	roundTimeout = 500 * time.Millisecond
+
+	defaultChanCapacity = 200
+)
+
+type Hashable interface {
+	HashString() string
+}
+
+type Signature struct {
+	Signer core.Address
+	Value  []byte
+}
+
+type RoundMessage[T Hashable] struct {
+	Value     *T
+	SessionId string
+
+	Round      int
+	Signatures []Signature
+}
+
+func (m RoundMessage[T]) Encode() []byte {
+	var buff bytes.Buffer
+
+	encoder := gob.NewEncoder(&buff)
+	_ = encoder.Encode(m)
+
+	return buff.Bytes()
+}
+
+func DecodeRoundMessage[T Hashable](data []byte) (RoundMessage[T], error) {
+	var msg RoundMessage[T]
+
+	decoder := gob.NewDecoder(bytes.NewReader(data))
+	if err := decoder.Decode(&msg); err != nil {
+		return RoundMessage[T]{}, errors.Wrap(err, "failed to decode round message")
+	}
+
+	return msg, nil
+}
+
+func (m RoundMessage[T]) SignatureValid(signature Signature) bool {
+	pubKey, err := crypto.SigToPub(m.SignHash(), signature.Value)
+	if err != nil {
+		return false
+	}
+	recovered := crypto.PubkeyToAddress(*pubKey)
+
+	return bytes.Equal(recovered.Bytes(), signature.Signer.Bytes())
+}
+
+func (m RoundMessage[T]) SignHash() []byte {
+	var buf bytes.Buffer
+
+	encoder := gob.NewEncoder(&buf)
+	_ = encoder.Encode(m.SessionId)
+	// gob cannot encode nil values, but they are valid
+	if m.Value != (*T)(nil) {
+		_ = encoder.Encode(m.Value)
+	}
+
+	for _, sig := range m.Signatures {
+		_ = encoder.Encode(sig.Signer)
+		_ = encoder.Encode(sig.Value)
+	}
+
+	hash := sha256.Sum256(buf.Bytes())
+	return hash[:]
+}
+
+type ReliableBroadcastMsg[T Hashable] struct {
+	Msg    RoundMessage[T]
+	Sender core.Address
+}
+
+// ReliableBroadcaster is a reliable broadcast protocol implementation based on the Dolev-Strong protocol.
+// It ensures that a same message is delivered to all parties in the group.
+// It is designed to work in a synchronous network with n > t,
+// where n is the number of parties and t is the maximum number of malicious parties.
+//
+// Instead of running the relay rounds one by one, it runs one big round and processes all incoming messages,
+// ensuring each early or late but valid message is processed.
+type ReliableBroadcaster[T Hashable] struct {
+	sessionId   string
+	parties     []p2p.Party
+	self        core.Account
+	logger      *logan.Entry
+	requestType p2p.RequestType
+
+	relayRounds int
+	broadcaster *Broadcaster
+	partiesMap  map[core.Address]bool
+
+	originMsgSender core.Address
+
+	// sender -> round -> received
+	receivedMsgs         map[core.Address]map[int]bool
+	values               map[string]*T
+	msgs                 chan ReliableBroadcastMsg[T]
+	finalSigChainReached bool
+}
+
+func NewReliable[T Hashable](
+	sessionId string,
+	parties []p2p.Party,
+	self core.Account,
+	threshold int,
+	requestType p2p.RequestType,
+	logger *logan.Entry,
+) *ReliableBroadcaster[T] {
+	// relay rounds are calculated as the t + 1,
```

Comment:
we used to name t a threshold in the tss ecdsa. maybe lets name the number of possible malicious parties as f